File: server/domain/planner/blueprint.py
"""The generated plan as plain frozen dataclasses — the shape before it is a row or JSON.

Mirrors `Plan -> Mesocycle -> Microcycle -> PlannedSession -> SessionBlock -> PrescribedSet`
field for field, with the departures listed below. Dataclasses rather than Pydantic so the
domain gains no dependency on the API layer; `server/plans/routes.py` maps these to response
models by hand, the same stitching `server/library/routes.py` does for the library.

## The five documented departures from the persisted tree

1. **No ids and no back-references.** The tree is nested, so parentage is structural.
   `Microcycle.plan_id`'s denormalisation and its composite foreign key are an insert-time
   concern for PR #11b, not something a preview can express.
2. **`exercise_key`, not `exercise_id`.** The domain is DB-free and speaks keys; the client
   already holds `key` from `useLibrary()`, and it saves a `SELECT`. #11b resolves keys to
   ids at persist time.
3. **`PlannedSession.status` omitted.** A preview has no status. #11b takes the column's
   `'planned'` default rather than sending one over the wire.
4. **⚠️ `rest_between_sets_seconds` has no home in the persisted tree.**
   `prescription_template` has it, but `session_block` has only `rest_after_seconds` (rest
   *after* the block) and `prescribed_set` only `target_rest_seconds` (rest *within* a set).
   The blueprint carries the template's value so the preview is honest about what it
   prescribes; **#11b needs a new column or a redefinition**, and taking that decision now
   beats discovering it mid-insert.
5. **`target_load_kg` / `target_grade_id` are present and always `None` in v1.0.0.** The
   wire shape stays stable for when they are filled, and deriving a load is the one place a
   bodyweight figure could creep into a prescription (CLAUDE.md, "never recommends losing
   weight").

## What `__post_init__` checks, and what it deliberately does not

Only a **schema CHECK**. A blueprint that could not be inserted is worth failing on in the
generator's own tests rather than in #11b's first bulk insert. `String(80)` limits are NOT
checked here: they are a column width, not a CHECK, and the safety guard already asserts
every generated string fits — duplicating it would turn that guard's red into a traceback
from a constructor, which is a worse failure to read.
"""

import enum
from dataclasses import dataclass
from datetime import date
from decimal import Decimal
from typing import Final

from server.domain.grades import Discipline
from server.domain.vocabulary import ActivityKind, Phase, ProtocolKind

# `ck_plan_week_count_in_range`. The upper bound is why `MAX_BLOCKS` exists at all.
MIN_WEEK_COUNT: Final = 1
MAX_WEEK_COUNT: Final = 52


class NoteKind(enum.StrEnum):
    """Why a plan carries a note. Closed, so the client can style or order them.

    A note is **never a gate**: the plan is complete and nothing is disabled. It exists so
    that a plan which is not quite what was asked for says so, in the plan, instead of
    leaving the user to notice.
    """

    FEWER_SESSIONS_THAN_REQUESTED = "fewer_sessions_than_requested"
    TARGET_BEYOND_ONE_PLAN = "target_beyond_one_plan"


@dataclass(frozen=True, slots=True)
class ScheduleNote:
    """One honest caveat about the plan as a whole. `kind` is the contract, `message` is copy."""

    kind: NoteKind
    message: str


@dataclass(frozen=True, slots=True)
class Shortfall:
    """An aspect this phase cannot train with the gear the climber has, and what would fix it.

    `options` is the minimal set of equipment combinations that would unlock the cell: each
    inner tuple is an AND set, and the outer tuple is the OR. It names **equipment rows
    only** — never a movement substitute, never an improvised edge (CLAUDE.md, the
    finger-loading safety boundary), which is why `substitution_hint` is not read here.

    Attached to the block it displaced and rolled up, deduped by `(phase, aspect_key)`, to
    the plan. Never a gate.
    """

    phase: Phase
    aspect_key: str
    options: tuple[tuple[str, ...], ...]
    message: str


@dataclass(frozen=True, slots=True)
class SetBlueprint:
    """One prescribed set. `prescribed_set` minus its ids."""

    set_index: int
    target_reps: int | None = None
    target_work_seconds: int | None = None
    target_rest_seconds: int | None = None
    target_intensity_pct: int | None = None
    target_rpe: int | None = None
    target_load_kg: Decimal | None = None
    target_grade_id: int | None = None

    def __post_init__(self) -> None:
        if self.set_index < 1:
            raise ValueError(
                f"set_index is 1-based (ck_prescribed_set_set_index_positive), "
                f"got {self.set_index}."
            )
        if self.target_rpe is not None and not 1 <= self.target_rpe <= 10:
            raise ValueError(
                f"target_rpe is 1-10 (ck_prescribed_set_target_rpe_in_range), "
                f"got {self.target_rpe}."
            )
        if self.target_intensity_pct is not None and not 1 <= self.target_intensity_pct <= 200:
            raise ValueError(
                f"target_intensity_pct is 1-200 (ck_prescribed_set_target_intensity_pct_sane), "
                f"got {self.target_intensity_pct}."
            )


@dataclass(frozen=True, slots=True)
class BlockBlueprint:
    """One exercise in one session. `protocol_kind` is snapshotted, as `session_block` does.

    `shortfall` is set when this block's aspect was **displaced** — the slot's intended
    aspect had no prescribable candidate, so the next emphasis entry that did was used and
    the miss is recorded on the block that replaced it.
    """

    order_index: int
    exercise_key: str
    aspect_key: str
    protocol_kind: ProtocolKind
    sets: tuple[SetBlueprint, ...]
    rest_after_seconds: int | None = None
    # Departure 4: carried so the preview is honest, with nowhere to persist it yet.
    rest_between_sets_seconds: int | None = None
    shortfall: Shortfall | None = None


@dataclass(frozen=True, slots=True)
class SessionBlueprint:
    """One prescribed session on one day. `weekday` is 0-6, Monday = 0.

    `blocks` may be empty in exactly one case, and it is documented rather than prevented:
    zero equipment with every injury area open leaves no safe candidate at all, so the
    session becomes `activity_kind=other` with a shortfall naming the injuries. The
    invariant is never an *unexplained* empty session.
    """

    weekday: int
    scheduled_on: date
    activity_kind: ActivityKind
    title: str
    estimated_minutes: int | None
    blocks: tuple[BlockBlueprint, ...]
    shortfalls: tuple[Shortfall, ...] = ()

    def __post_init__(self) -> None:
        if not 0 <= self.weekday <= 6:
            raise ValueError(
                f"weekday is 0-6 with Monday = 0 (ck_planned_session_weekday_in_range), "
                f"got {self.weekday}."
            )
        if self.scheduled_on.weekday() != self.weekday:
            raise ValueError(
                f"scheduled_on {self.scheduled_on} is a {self.scheduled_on.strftime('%A')} "
                f"but weekday says {self.weekday}. `planned_session` stores both and nothing "
                f"in the schema keeps them in agreement, so the generator has to."
            )


@dataclass(frozen=True, slots=True)
class MicrocycleBlueprint:
    """One week. `phase` is the parent mesocycle's, denormalised the way the read path wants it.

    `is_deload` is exactly `phase is Phase.DELOAD` and nothing cleverer — a taper is
    identified by its phase, not by this flag, because a taper is not a deload.
    """

    week_no: int
    start_date: date
    is_deload: bool
    phase: Phase
    sessions: tuple[SessionBlueprint, ...]

    def __post_init__(self) -> None:
        if self.week_no < 1:
            raise ValueError(
                f"week_no is 1-based (ck_microcycle_week_no_positive), got {self.week_no}."
            )
        if self.start_date.weekday() != 0:
            raise ValueError(f"a microcycle starts on a Monday; {self.start_date} does not.")


@dataclass(frozen=True, slots=True)
class MesocycleBlueprint:
    """A phase block. `start_week` / `end_week` are 1-based and inclusive, as the column is."""

    phase: Phase
    start_week: int
    end_week: int
    microcycles: tuple[MicrocycleBlueprint, ...]

    def __post_init__(self) -> None:
        if self.start_week < 1:
            raise ValueError(
                f"start_week is 1-based (ck_mesocycle_start_week_positive), got {self.start_week}."
            )
        if self.end_week < self.start_week:
            raise ValueError(
                f"end_week {self.end_week} precedes start_week {self.start_week} "
                f"(ck_mesocycle_end_week_after_start)."
            )


@dataclass(frozen=True, slots=True)
class PlanBlueprint:
    """A whole generated plan. The root the endpoint serialises and #11b inserts.

    `shortfalls` is the plan-level roll-up, deduped by `(phase, aspect_key)` from the blocks;
    `notes` is everything the plan wants to say about itself. Both may be empty, and an empty
    one is the normal case for a well-equipped climber with enough days.
    """

    name: str
    discipline: Discipline
    target_grade_id: int | None
    current_grade_id: int | None
    start_date: date
    week_count: int
    grade_gap: int
    mesocycles: tuple[MesocycleBlueprint, ...]
    shortfalls: tuple[Shortfall, ...] = ()
    notes: tuple[ScheduleNote, ...] = ()

    def __post_init__(self) -> None:
        if not MIN_WEEK_COUNT <= self.week_count <= MAX_WEEK_COUNT:
            raise ValueError(
                f"week_count must be {MIN_WEEK_COUNT}-{MAX_WEEK_COUNT} "
                f"(ck_plan_week_count_in_range), got {self.week_count}."
            )
        if self.start_date.weekday() != 0:
            raise ValueError(f"a plan starts on a Monday; {self.start_date} does not.")

File: web/src/plan/blueprint.ts
import type {
  LibraryExercise,
  Phase,
  PlanBlock,
  PlanSession,
  Prescription,
  Profile,
} from '../api/types';
import { humanise, prescriptionLine } from '../library/browse';

/**
 * The pure half of the `/plan` screen: what may be asked for, what identifies the answer, and
 * the joins and one-line summaries the markup needs. No React, no fetch, no rendering — so the
 * decisions below are testable without mounting anything.
 *
 * ## Why the client decides whether to ask at all
 *
 * `POST /api/plans/preview` answers an unplannable profile with a 422 carrying one fixed
 * sentence, and that stays as defence in depth. But the client already holds the profile, so it
 * can see the refusal coming and show the sentence — plus a way to fix it — without spending a
 * request, a Neon wake and a 422 on a question it could already answer. That is what
 * `previewBlocker` is for, and it is why `usePlanPreview` is `enabled` on it.
 *
 * ⚠️ **The five sentences below are copied VERBATIM from
 * `server/domain/planner/contract.py::REFUSAL_MESSAGES`**, reworked on Kilian's dev-server
 * sign-off (2026-08-24) and frozen. Do not reword, retitle or repunctuate one. They are
 * duplicated rather than fetched because the whole point is to have them before the request —
 * and the duplication is self-checking: `tests/test_planner_refusal_copy.py` parses this
 * declaration and fails until both copies say the same thing.
 *
 * The sixth refusal, `cross_discipline_grades`, is deliberately **not** here: spotting it needs
 * the grade ladder from `GET /api/vocabulary`, both pickers are locked to one scale so it should
 * be unreachable, and if it does happen the server's own 422 says so. Splitting it that way
 * keeps this module free of the vocabulary.
 */

/** Which answer is missing. Mirrors `RefusalReason`'s values for the five cases visible here. */
export type PreviewBlockerReason =
  | 'no_target_grade'
  | 'no_current_grade'
  | 'sessions_per_week_unanswered'
  | 'available_weekdays_unanswered'
  | 'no_available_days';

export interface PreviewBlocker {
  reason: PreviewBlockerReason;
  message: string;
  /**
   * Where the answer is given. `/onboarding` resumes at the first unanswered step — but it
   * redirects a COMPLETE profile to `/profile`, and `available_weekdays === 0` is a complete
   * answer (`profile/completion.ts` counts the step done), so that one case has to go straight
   * to the editor or the link bounces.
   */
  fix: '/onboarding' | '/profile';
}

const BLOCKER_MESSAGES: Record<PreviewBlockerReason, string> = {
  no_target_grade:
    "Your plan is built around a target grade. Finish setting up your profile and we'll build it.",
  no_current_grade:
    "Your plan needs to know what you climb now as well as what you're aiming at. Finish " +
    "setting up your profile and we'll build it.",
  sessions_per_week_unanswered:
    "Your plan needs to know how often you can train. Finish setting up your profile and we'll " +
    'build it.',
  available_weekdays_unanswered:
    "Your plan needs to know which days you can train. Finish setting up your profile and we'll " +
    'build it.',
  no_available_days:
    "You haven't marked any day as available, so there's nowhere to put a session. Tick at " +
    "least one day in your profile and we'll build it.",
};

/**
 * The first answer the generator would refuse over, or `null` when it would build a plan.
 *
 * The order matches `server/plans/routes.py::_planner_input` exactly, so a profile missing two
 * answers names the same one on both sides of the wire. ⚠️ `available_weekdays` is tested
 * `=== null` before `=== 0`: zero is a legal *answer* ("answered, no days") with its own
 * sentence, and a truthiness check there would send the user to the wrong step.
 */
export function previewBlocker(profile: Profile): PreviewBlocker | null {
  const blocker = (reason: PreviewBlockerReason, fix: PreviewBlocker['fix']): PreviewBlocker => ({
    reason,
    message: BLOCKER_MESSAGES[reason],
    fix,
  });

  if (unanswered(profile.target_grade_id) || unanswered(profile.primary_discipline)) {
    return blocker('no_target_grade', '/onboarding');
  }
  if (unanswered(profile.current_grade_id)) return blocker('no_current_grade', '/onboarding');
  if (unanswered(profile.sessions_per_week)) {
    return blocker('sessions_per_week_unanswered', '/onboarding');
  }
  if (unanswered(profile.available_weekdays)) {
    return blocker('available_weekdays_unanswered', '/onboarding');
  }
  if (profile.available_weekdays === 0) return blocker('no_available_days', '/profile');
  return null;
}

/**
 * `null`, or absent altogether.
 *
 * Every field of `ProfileResponse` is required on the wire, so `undefined` should be
 * unreachable — but the two readings of an absent field are "unanswered" and "answered", and
 * only the first is safe: the second sends a profile the generator cannot use to the endpoint
 * and, downstream of that, reads `injuries` off a body that has none. Erring towards a notice
 * costs one sentence; erring the other way replaced the whole route with an error boundary in
 * `routeGuard.test.tsx`, whose fetch mock answers every request with a token payload.
 */
function unanswered(value: number | string | null | undefined): boolean {
  return value === null || value === undefined;
}

export function canPreview(profile: Profile | undefined): boolean {
  return profile !== undefined && previewBlocker(profile) === null;
}

/**
 * Everything the answer depends on, and nothing else — the query key.
 *
 * A profile change therefore yields a NEW key and a new fetch with no invalidation wiring, going
 * back to the screen is a cache hit, and `staleTime: Infinity` is correct rather than a
 * staleness bet: the same inputs against the same deploy cannot produce a different plan
 * (`server/models.py::Plan` promises exactly that, and the deploy pins the library).
 *
 * ⚠️ An injury's `note` and `started_on` are deliberately OUT. The planner reads open injury
 * *keys* and nothing else, so editing a note must not throw away a 32-week plan and pay for
 * another one. Ids, sorted, so the order the endpoint happened to return them in cannot fork the
 * key.
 */
export interface PreviewKeyParts {
  start_date: string;
  discipline: string | null;
  target_grade_id: number | null;
  current_grade_id: number | null;
  sessions_per_week: number | null;
  available_weekdays: number | null;
  strength_aspect_id: number | null;
  weakness_aspect_id: number | null;
  open_injuries: string;
}

export function previewKeyParts(profile: Profile, startDate: string): PreviewKeyParts {
  return {
    start_date: startDate,
    discipline: profile.primary_discipline,
    target_grade_id: profile.target_grade_id,
    current_grade_id: profile.current_grade_id,
    sessions_per_week: profile.sessions_per_week,
    available_weekdays: profile.available_weekdays,
    strength_aspect_id: profile.strength_aspect_id,
    weakness_aspect_id: profile.weakness_aspect_id,
    open_injuries: profile.injuries
      .map((injury) => injury.injury_area_id)
      .sort((a, b) => a - b)
      .join(','),
  };
}

/**
 * The Monday on or after `now`, as `YYYY-MM-DD` in the browser's own timezone.
 *
 * The client owns this because the domain has no clock and the server has no timezone: whatever
 * is sent is normalised to the Monday on or after it, so a local Monday round-trips unchanged.
 * Built from local date parts rather than `toISOString()`, which would shift the day by the
 * UTC offset and hand a Sunday to a planner that refuses anything but a Monday.
 */
export function nextMonday(now: Date): string {
  const shift = (8 - now.getDay()) % 7;
  const monday = new Date(now.getFullYear(), now.getMonth(), now.getDate() + shift);
  const month = String(monday.getMonth() + 1).padStart(2, '0');
  const day = String(monday.getDate()).padStart(2, '0');
  return `${String(monday.getFullYear())}-${month}-${day}`;
}

/**
 * `exercise_key -> exercise`, built ONCE for the screen.
 *
 * A block names a key, not an id (the domain is DB-free and speaks keys), so the join happens
 * here against the library the app has already fetched. **Built once and passed down**: the
 * worst-case plan is 224 sessions and 672 blocks, and rebuilding an 85-entry index per row is
 * the one thing on this screen that would be quadratic.
 *
 * An unknown key is simply absent — `namesOf`'s rule, and for the same reason: the only way to
 * reach one is a plan generated against a library newer than the client's. What `exerciseLabel`
 * does with that absence is a separate decision; see it.
 */
export function exercisesByKey(
  exercises: readonly LibraryExercise[],
): ReadonlyMap<string, LibraryExercise> {
  return new Map(exercises.map((exercise) => [exercise.key, exercise]));
}

/**
 * What to call a block's exercise.
 *
 * ⚠️ **Deliberately NOT `namesOf`'s drop.** An equipment id with no vocabulary row renders as a
 * bare integer, which tells the reader nothing, so dropping it is right. A key does not: it is
 * authored English, and `weighted_max_hangs` humanises to something a climber can act on. So a
 * block whose exercise this client has never heard of is still shown, with its prescribed sets,
 * under a slightly plainer name — hiding prescribed work would be the worse failure.
 */
export function exerciseLabel(key: string, index: ReadonlyMap<string, LibraryExercise>): string {
  return index.get(key)?.name ?? humanise(key);
}

/** Monday first, matching `available_weekdays` bit 0 and `planned_session.weekday`. */
const WEEKDAY_NAMES = [
  'Monday',
  'Tuesday',
  'Wednesday',
  'Thursday',
  'Friday',
  'Saturday',
  'Sunday',
] as const;

export function weekdayName(weekday: number): string {
  return WEEKDAY_NAMES[weekday] ?? 'Unscheduled';
}

export const MONTH_NAMES = [
  'Jan',
  'Feb',
  'Mar',
  'Apr',
  'May',
  'Jun',
  'Jul',
  'Aug',
  'Sep',
  'Oct',
  'Nov',
  'Dec',
] as const;

/**
 * `2026-08-31` -> `31 Aug 2026`, from the string's own parts.
 *
 * ⚠️ Not `new Date(iso).toLocaleDateString()`: a date-only string parses as UTC midnight, which
 * renders as the PREVIOUS day everywhere west of Greenwich. A plan whose first week starts on
 * Sunday is not a rendering nit — every session on the screen is then a day out.
 */
export function formatDay(iso: string): string {
  const [year, month, day] = iso.split('-');
  const name = month === undefined ? undefined : MONTH_NAMES[Number(month) - 1];
  if (year === undefined || day === undefined || name === undefined) return iso;
  return `${String(Number(day))} ${name} ${year}`;
}

/** A session's shape at a glance, for the disclosure summary. */
export function sessionSummary(session: PlanSession): string {
  const terms: string[] = [];
  const blocks = session.blocks.length;

  if (blocks === 0) {
    // The Recovery slot: zero blocks, and it carries a shortfall saying why. "0 blocks" would
    // read as a data fault; the invariant is that no empty session is UNEXPLAINED, not that
    // none exists.
    terms.push('nothing prescribed');
  } else {
    const sets = session.blocks.reduce((total, block) => total + block.sets.length, 0);
    terms.push(`${String(blocks)} ${blocks === 1 ? 'block' : 'blocks'}`);
    terms.push(`${String(sets)} ${sets === 1 ? 'set' : 'sets'}`);
  }
  // `null` for a session that prescribes nothing — no estimate, rather than an estimate of the
  // warm-up alone.
  if (session.estimated_minutes !== null) terms.push(`~${String(session.estimated_minutes)} min`);

  return terms.join(' · ');
}

/**
 * One block's prescription as a line, reusing `prescriptionLine`'s null-omission rule.
 *
 * ⚠️ **One line per BLOCK, not a row per set, and that is a measured decision.** In v1.0.0 every
 * set in a block is the same row of `prescription_template` with a different `set_index`, so a
 * row each would repeat itself — and the worst-case plan holds **2,421** of them. Reading the
 * first set and the count says exactly as much for a fraction of the DOM. The day sets stop
 * being identical (a per-set progression), this is the function that has to change.
 *
 * `phase` is the microcycle's, and it is here only because `Prescription` carries one;
 * `prescriptionLine` never reads it.
 */
export function setsLine(block: PlanBlock, phase: Phase): string {
  const first = block.sets[0];
  if (first === undefined) return '';

  const prescription: Prescription = {
    phase,
    sets: block.sets.length,
    reps: first.target_reps,
    work_seconds: first.target_work_seconds,
    rest_seconds: first.target_rest_seconds,
    rest_between_sets_seconds: block.rest_between_sets_seconds,
    intensity_pct: first.target_intensity_pct,
    target_rpe: first.target_rpe,
  };
  return prescriptionLine(prescription);
}
